### PR TITLE
pmap speedup

### PR DESCRIPTION
**Pull request by amitmurthy**
Partially addresses #17301 

On my machine:

```
x = [bitrand(4,2) for i = 1:100000];   
@time pmap(identity, x);
```

this PR takes 8.8 seconds vs 6.6 seconds on 0.4

### CODE REVIEW

File: base/asyncmap.jl [hunk 1]
```diff
@@ -19,44 +19,97 @@ type AsyncCollector
...
 function AsyncCollector(f, results, c...; ntasks=0)
     if ntasks == 0
         ntasks = max(nworkers(), 100)
+        max_tasks = ()->ntasks
+    elseif isa(ntasks, Integer)
+        max_tasks = ()->ntasks
+    elseif isa(ntasks, Function)
+        max_tasks = ntasks
+    else
+        throw(ArgumentError("ntasks must be an Integer or a zero-arg function returning the maximum number of tasks allowed."))
```
Comment (tkelman) [inline base/asyncmap.jl]: why is a zero-arg function desirable for this?

File: test/parallel_exec.jl [hunk 2]
```diff
@@ -801,7 +805,7 @@ end
 @test [1:100...] == pmap(x->x, Base.Generator(x->(sleep(0.0001); x), 1:100))
 
 # Test asyncmap
-@test allunique(asyncmap(x->object_id(current_task()), 1:100))
+@test allunique(asyncmap(x->(sleep(1.0);object_id(current_task())), 1:10))
```
Comment (tkelman) [inline test/parallel_exec.jl]: what's the reason for these changes?

File: base/asyncmap.jl (hunk 1 above)
Comment (amitmurthy) [inline base/asyncmap.jl]: Allows for the number of concurrent tasks in `AsyncCollector` to be dynamic. One of the requested features was that pmap must be able to leverage workers as and when they come online. This allows for a user of `AsyncCollector `, like `pmap` to increase count of parallel tasks depending on current number of workers. Note that it is not a major issue if the number of tasks is more than the number of workers, but it will be quite inefficient the other way around.

File: test/parallel_exec.jl (hunk 2 above)
Comment (amitmurthy) [inline test/parallel_exec.jl]: Previously each item was being executed in a new task. Now the same set of tasks run through the entire collection. The `sleep` is required as otherwise the first task itself will map the entire range (as the computation is non-yielding)

File: base/asyncmap.jl (hunk 1 above)
Comment (tkelman) [inline base/asyncmap.jl]: so this would be accomplished via a zero-arg function that reads from some global runtime state?

File: test/parallel_exec.jl (hunk 2 above)
Comment (tkelman) [inline test/parallel_exec.jl]: if this would fail without this commit, then it should be squashed

File: base/asyncmap.jl (hunk 1 above)
Comment (amitmurthy) [inline base/asyncmap.jl]: A `WorkerPool` allows for workers to be dynamically added. `pmap` now passes this keyword arg as `ntasks=()->nworkers(p)` where `p` is the worker pool.

File: test/parallel_exec.jl (hunk 2 above)
Comment (amitmurthy) [inline test/parallel_exec.jl]: reordered and squashed.

File: base/channels.jl [hunk 3]
```diff
@@ -51,7 +51,6 @@ function fetch(c::Channel)
 end
 
 function take!(c::Channel)
-    !isopen(c) && !isready(c) && throw(closed_exception())
```
Comment (tkelman) [inline base/channels.jl]: does the added test in the next commit exercise this?

File: base/asyncmap.jl [hunk 4]
```diff
@@ -19,44 +19,97 @@ type AsyncCollector
     f
     results
     enumerator::Enumerate
-    ntasks::Int
+    max_tasks::Function
+    c::Channel{Tuple{Int, Any}}     # to communicate with the tasks
```
Comment (tkelman) [inline base/asyncmap.jl]: a slightly more descriptive field name than `c` might be good for clarity's sake

File: base/channels.jl (hunk 3 above)
Comment (amitmurthy) [inline base/channels.jl]: Well yes, because that is how I found the bug. But it ought to have an explicit test anyway. Will add one.
